services/whatpulse/whatpulse.service.js

```diff
@@ -0,0 +1,128 @@
+import Joi from 'joi'
+import { BaseJsonService, NotFound } from '../index.js'
+import { ordinalNumber } from '../text-formatters.js'
+
+const schema = Joi.object({
+  Keys: Joi.alternatives(Joi.string(), Joi.number()),
+  Clicks: Joi.alternatives(Joi.string(), Joi.number()),
+  UptimeShort: Joi.string().required(),
+  Download: Joi.string().required(),
+  Upload: Joi.string().required(),
+  Ranks: Joi.object({
+    Keys: Joi.string().required(),
+    Clicks: Joi.string().required(),
+    Download: Joi.string().required(),
+    Upload: Joi.string().required(),
+    Uptime: Joi.string().required(),
+  }),
+}).required()
+
+const queryParamSchema = Joi.object({
+  category: Joi.string().required(),
+}).required()
+
+export default class WhatPulse extends BaseJsonService {
+  static category = 'activity'
+  static route = {
+    base: 'whatpulse',
+    pattern: ':userOrTeam/:id',
+    queryParamSchema,
+  }
```

Comment:
There's a few things to bring up with the URLs here. Docs for this are at https://github.com/badges/shields/blob/master/doc/badge-urls.md

- If a param can only be one of a closed list of options, we can specify that in the URL definition and it both saves us some manual validation (because validation is enforced at the router level) and the frontend renders it as a drop-down in the builder UI. So instead of `:userOrTeam` we could make that `:userType(user|team)`. Example of this:


![Screenshot at 2022-10-03 16-46-26](https://user-images.githubusercontent.com/6025893/193623644-301b7b95-bd86-46a2-a7c1-1d2c0305c745.png)

- Category is a required param, so it should be in the main body of the route, not a query param
- When we allow the user to pick the data point or statistic we're going to report, we usually call this param `:metric` and again restrict it to a closed list e.g: `:metric(keys|clicks|uptime|download|upload)`. It makes it easy to see what the options are and again does some validation for us. If the route matched, we know we have one of those options.
- Rather than smooshing the metric we're going to display and whether we want to view it as a rank or the raw value into one param that we then have to awkwardly parse ourselves, lets separate those concepts. Referring again to our docs on URLs https://github.com/badges/shields/blob/master/doc/badge-urls.md it would seem reasonable to default to metric and have an optional query param `?rank`

So.. putting all that together, that gives us something like:

```js
route = {
  base: 'whatpulse',
  pattern: ':metric(keys|clicks|uptime|download|upload)/:userType(user|team)/:id',
}
```

with a `queryParamSchema` that can accept an optional param `rank` (Something with a `include_prereleases` flag is probably a good place to crib from on this).

Doing that should bring things in line with the conventions, simplify the `transform()` method a bit, reduce the number of examples we need to provide and make it clearer to a user what the options are.

Reply:
Very cool suggestions, thanks!
I had no clue about the possible dropdown usage in the badge's modal, nice :)

You're right, `metric` is the perfect word here + `rank` looks well as a query param.

Updated.